### PR TITLE
Add app template supported AWS lambda s3 event-based predictions

### LINKED ISSUE #19: AWS Lambda event-based s3 event serving support
The purpose of this task is to add a `unionml init` template that creates a unionml project that supports model serving endpoints that use s3 events that invoke the prediction endpoint for generating predictions based on dumping a file onto s3.

#### Resources

- https://egghead.io/lessons/aws-deploy-an-s3-bucket-and-trigger-an-aws-lambda-function-with-aws-sam
- https://docs.aws.amazon.com/lambda/latest/dg/with-s3-example.html

### PR DESCRIPTION
**Pull request by cosmicBboy**
Fixes #19 

This PR creates a unionml app template that can be instantiated with `unionml init --template basic-aws-lambda-s3 APP_NAME`, which ships with the necessary SAM template, app code, handler code, and basic unit tests for users to get started with s3-event-based predictions.

### CODE REVIEW

File: tests/unit/aws_lambda_app/s3_event.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+        # generate prediction
+        predictions = model.predict(features=features)
+        logger.info(f"generated predictions {predictions}")
+
+        # upload predictions to s3
+        out_filename = "/tmp/predictions.json"
+        with open(out_filename, "w") as f:
+            json.dump(predictions, f)
+
```
Comment (eapolinario) [inline tests/unit/aws_lambda_app/s3_event.py:38]: any reason not to use another `NamedTemporaryFile` for the predictions as well?

File: tests/unit/test_aws_lambda_handler.py [hunk 2]
```diff
@@ -80,3 +125,34 @@ def test_lambda_handler(monkeypatch, apigw_event, tmpdir):
...
+        ...
+
+    with patch.object(s3_client, "download_file", MagicMock(wraps=mock_download_file)) as mock_dl_file, patch.object(
+        s3_client, "upload_file", MagicMock(wraps=mock_upload_file)
+    ) as mock_ul_file:
+        ret = lambda_handler(s3_event, "")
+
+    assert ret is None
+    mock_dl_file.assert_called_once()
+    mock_ul_file.assert_called_once()
```
Comment (eapolinario) [inline tests/unit/test_aws_lambda_handler.py:158]: is there anyway we could capture the content of the file being written to s3? Just to double-check we're writing the right predictions.

**Review by eapolinario — COMMENTED** [review]
looking good. Left 2 comments, nothing major.

File: tests/unit/aws_lambda_app/s3_event.py (hunk 1 above)
Comment (cosmicBboy) [reply tests/unit/aws_lambda_app/s3_event.py:38]: it somehow wasn't working, the file uploaded to s3 would be empty 🤷‍♂️

File: tests/unit/test_aws_lambda_handler.py (hunk 2 above)
Comment (cosmicBboy) [reply tests/unit/test_aws_lambda_handler.py:158]: lemme try something

File: unionml/model.py [hunk 3]
```diff
@@ -515,15 +515,26 @@ def save(self, file: Union[str, os.PathLike, IO], *args, **kwargs):
...
+        """
         self.artifact = ModelArtifact(self._loader(file, *args, **kwargs))
         return self.artifact.model_object
 
-    def load_from_env(self, env_var: str = "UNIONML_MODEL_PATH"):
-        """Load a model object from an environment variable pointing to the model file."""
+    def load_from_env(self, env_var: str = "UNIONML_MODEL_PATH", *args, **kwargs):
+        """Load a model object from an environment variable pointing to the model file.
+
+        :env_var: environment variable referencing a path to load a model from.
```
Comment (eapolinario) [inline unionml/model.py:530]: nit: missing `param` before `env_var`.

File: unionml/templates/basic-aws-lambda-s3/{{cookiecutter.app_name}}/app.py [hunk 4]
```diff
@@ -69,13 +69,12 @@ def lambda_handler(event, context):
...
-        with open(out_filename, "w") as f:
-            json.dump(predictions, f)
-
-        upload_key = f"predictions/{key.split('/')[-1]}"
-        s3_client.upload_file(out_filename, bucket, upload_key)
-        logger.info(f"uploaded predictions to {bucket}/{upload_key}")
+        with tempfile.NamedTemporaryFile("w") as out_file:
+            json.dump(predictions, out_file)
+            upload_key = f"predictions/{key.split('/')[-1]}"
+            out_file.flush()
```
Comment (eapolinario) [inline unionml/templates/basic-aws-lambda-s3/{{cookiecutter.app_name}}/app.py:75]: nice.

**Review by eapolinario — APPROVED** [review]

File: docs/source/reacting_to_s3_events.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
+(reacting_to_s3_events)=
+
+# Reacting to S3 Events
+
+In {ref}`serving_aws_lambda` we learned how to deploy a prediction service to an AWS
```
[the remark this answers is not in the document]
Comment (cosmicBboy) [reply docs/source/reacting_to_s3_events.md:5]: actually using https://myst-parser.readthedocs.io/en/latest/ for these docs... we should move Flyte (incrementally) to this, it's a lot easier to write docs for

**Pull request merged**